pkg/kubelet/volume/plugins.go

```diff
@@ -0,0 +1,177 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package volume
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/types"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/golang/glog"
+)
+
+// Plugin is an interface to volume plugins.
+type Plugin interface {
+	// Init initializes the plugin.  This will be called exactly once
+	// before any New* calls are made - implementations of plugins may
+	// assert this.
+	Init(host Host)
+
+	// Name returns the plugin's name.  Plugins should use namespaced names
+	// such as "example.com/volume".  The "kubernetes.io" namespace is
+	// reserved for plugins which are bundled with kubernetes.
+	Name() string
+
+	// CanSupport tests whether the Plugin supports a given volume
+	// specification from the API.
+	CanSupport(spec *api.Volume) bool
+
+	// NewBuilder creates a new volume.Builder from an API specification.
+	// - spec: The api.Volume spec
+	// - podUID: The UID of the enclosing pod
+	NewBuilder(spec *api.Volume, podUID types.UID) (Builder, error)
+
+	// NewCleaner creates a new volume.Cleaner from recoverable state.
+	// - name: The volume name, as per the api.Volume spec.
+	// - podUID: The UID of the enclosing pod
+	NewCleaner(name string, podUID types.UID) (Cleaner, error)
+}
+
+// Host is an interface that plugins can use to access the kubelet.
+type Host interface {
+	// GetRootDir returns the absolute path of the kubelet's data directory.
+	GetRootDir() string
+
+	// GetPluginsDir returns the absolute path to the directory under which
+	// plugins may store data.  Plugins should make subdirs of this named
```

Comment:
OK, I retooled all of this.  The exposed-to-plugins interface now does all the path joining and returns the full path name.  Plugins can do what they want below that.

I also made all of the get*Dir() methods private and added a purpose-built volumeHost type to pass to plugins.

I opted to make these return paths, rather than mkdir, mostly because it was a smaller net change.  I think it doesn't matter much, but if someone wants to call me out on it, I could retool that part again.